core/client/src/main/java/alluxio/client/file/policy/DeterministicHashPolicy.java

```diff
@@ -0,0 +1,128 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file.policy;
+
+import alluxio.client.block.BlockWorkerInfo;
+import alluxio.wire.WorkerNetAddress;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Random;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This policy maps blockId to several deterministic Alluxio workers. The number of workers a block
+ * can be mapped to can be passed through the constructor. The default is 1. It skips the workers
+ * that do not have enough capacity to hold the block.
+ */
+@NotThreadSafe
+public final class DeterministicHashPolicy implements BlockLocationPolicy {
+  /** The default number of shards to serve a block. */
+  private static final int DEFAULT_NUM_SHARDS = 1;
+  private final int mShards;
+  private final Random mRandom = new Random();
+  private List<BlockWorkerInfo> mWorkerInfoList;
+  private boolean mInitialized = false;
+
+  /**
+   * Constructs a new {@link DeterministicHashPolicy}.
+   */
+  public DeterministicHashPolicy() {
+    this(DEFAULT_NUM_SHARDS);
+  }
+
+  /**
+   * Constructs a new {@link DeterministicHashPolicy}.
+   *
+   * @param numShards the number of shards a block's traffic can be sharded to
+   */
+  public DeterministicHashPolicy(Integer numShards) {
+    Preconditions.checkArgument(numShards >= 1);
+    mShards = numShards;
+  }
+
+  @Override
+  public WorkerNetAddress getWorkerForBlock(Iterable<BlockWorkerInfo> workerInfoList,
+      long blockId, long blockSizeBytes) {
+    if (!mInitialized) {
+      mWorkerInfoList = Lists.newArrayList(workerInfoList);
+      Collections.sort(mWorkerInfoList, new Comparator<BlockWorkerInfo>() {
+        @Override
+        public int compare(BlockWorkerInfo o1, BlockWorkerInfo o2) {
+          return o1.getNetAddress().toString().compareToIgnoreCase(o2.getNetAddress().toString());
+        }
+      });
+      mInitialized = true;
+    }
+
+    // The worker information (e.g. capacity) in the blockWorkerInfoMap can be more up-to-date
+    // than those in mWorkerInfoList. We assume that the worker information is more likely to
+    // change than the worker network address.
+    HashMap<WorkerNetAddress, BlockWorkerInfo> blockWorkerInfoMap = new HashMap<>();
+    for (BlockWorkerInfo workerInfo : workerInfoList) {
+      blockWorkerInfoMap.put(workerInfo.getNetAddress(), workerInfo);
+    }
+
+    List<WorkerNetAddress> workers = new ArrayList<>();
+    // Try the next one if the worker mapped from the blockId doesn't work until all the workers
+    // are examined.
+    int index = (int) (blockId % (long) mWorkerInfoList.size());
```

Comment:
I think the idea is to assume that new workers won't be used. I followed the style in RoundRobinPolicy. I personally prefer to always using mWorkerInfoList since it is unlikely to have capacity changes during a file read/write.  I am also fine with using workerInfoList all the time with the cost of sorting for every block. but that might be a little cpu-expensive if the number of workers is huge (say 100k) :) 

what do you prefer?